bare-metal-fractal/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+extern crate wee_alloc;
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use core::f32;
+use core::ops::Add;
+
+// Compiler calming
+#[panic_handler]
+fn handle_panic(_: &core::panic::PanicInfo) -> ! {
+    loop {}
+}
+
+#[alloc_error_handler]
+fn error_handler(_: core::alloc::Layout) -> ! {
+    loop {}
+}
+
+#[lang = "eh_personality"]
+extern "C" fn eh_personality() {}
+
+extern "C" {
+    fn js_logu(x: u32);
+    fn js_logf(x: f32);
+}
+
+// Memory Management
+#[repr(C)]
+pub struct TileBuffer {
+    w: u32,
+    h: u32,
+    buf: Vec<u32>,
+}
+
+// http://jakegoulding.com/rust-ffi-omnibus/objects/
+#[no_mangle]
+pub extern "C" fn alloc_tile(width: u32, height: u32) -> *mut TileBuffer {
+    let tile = TileBuffer {
+        w: width,
+        h: height,
+        buf: Vec::with_capacity((width * height) as usize),
+    };
+    Box::into_raw(Box::new(tile))
+}
+
+#[no_mangle]
+pub extern "C" fn get_buffer(tile_ptr: *mut TileBuffer) -> *mut u32 {
+    let tile = unsafe {
+        assert!(!tile_ptr.is_null());
+        &mut *tile_ptr
+    };
+
+    tile.buf.as_mut_ptr()
+}
+
+#[no_mangle]
+pub extern "C" fn free_tile(tile_ptr: *mut TileBuffer) {
+    if tile_ptr.is_null() {
+        return;
+    }
+    unsafe {
+        Box::from_raw(tile_ptr);
+    }
+}
+
+// Complex coordination
+// https://rustwasm.github.io/wasm-bindgen/examples/julia.html
+#[derive(Clone, Copy)]
+struct Complex {
+    re: f64,
+    im: f64,
+}
+
+impl Complex {
+    fn square(self) -> Complex {
+        let re = (self.re * self.re) - (self.im * self.im);
+        let im = 2.0 * self.re * self.im;
+        Complex { re, im }
+    }
+
+    fn norm(&self) -> f64 {
+        (self.re * self.re) + (self.im * self.im)
+    }
+}
+
+impl Add<Complex> for Complex {
+    type Output = Complex;
+
+    fn add(self, rhs: Complex) -> Complex {
+        Complex {
+            re: self.re + rhs.re,
+            im: self.im + rhs.im,
+        }
+    }
+}
+
+// Mandelbrot maths
+fn mandel_iter(max_iter: u64, c: Complex) -> u64 {
+    let mut z: Complex = c;
+
+    let mut iter = 1;
+
+    while (z.norm() <= 4.0) && (iter < max_iter) {
+        z = c + z.square();
+        iter = iter + 1;
+    }
+
+    if iter == max_iter {
+        0
+    } else {
+        iter
+    }
+}
+
+const fn rgb(r: u8, g: u8, b: u8) -> u32 {
+    255 << 24 | ((b as u32) << 16) | ((g as u32) << 8) | (r as u32)
+}
+
+fn rgbf(r: f32, g: f32, b: f32) -> u32 {
+    rgb((r * 255.0) as u8, (g * 255.0) as u8, (b * 255.0) as u8)
+}
+
+fn scale_value(value: f32, v_low: f32, v_high: f32, s_low: f32, s_high: f32) -> f32 {
+    (((value - v_low) / (v_high - v_low)) * (s_high - s_low)) + s_low
+}
+
+fn hsv_to_rgb(h: f32, s: f32, v: f32) -> u32 {
+    let i = (h * 6.0) as u8 as f32;
+    let f = h * 6.0 - i;
+    let p = v * (1.0 - s);
+    let q = v * (1.0 - f * s);
+    let t = v * (1.0 - (1.0 - f) * s);
+    if h < 0.16666666666 {
```

Comment:
Might look better written as

```rust
match h {
  h if h < 0.16666666 => rgbf(q, v, p),
  h if // etc
}
```

Additionally, I suspect we save CPU cycles if we inline the variables into the corresponding branches. Otherwise we compute them all every time when only one branch will actually execute

Reply:
This code actually added a substantial number of bytes due to constants, and runs a grand total of PALETTE_SIZE times (the palette is precomputed into an array)
Formatting into a match left the size unchanged, but looks a bit prettier